### PR TITLE
Adding documents to TE/JAX

### PR DESCRIPTION
**Pull request by mingxu1067**

### CODE REVIEW

**Comment by ksivaman** [comment]
@mingxu1067 @jeng1220 Looks like there are a lot of duplicate commits here, could you please rebase with main?

**Comment by ksivaman** [comment]
Since #54 is merged now, we can remove the WIP tag too

**Comment by jeng1220** [comment]
> @mingxu1067 @jeng1220 Looks like there are a lot of duplicate commits here, could you please rebase with main?

@mingxu1067 ,
Could you update your branch first? So other colleagues can read the change easier and earlier.

**Comment by ksivaman** [comment]
Could we combine #88 with this PR? I think the scope of this PR is documentation and that change falls within this. I want to avoid burdening the ci with a bunch of small PRs especially when they can be combined. This would speed up the dev/review/merge process. @mingxu1067

**Comment by mingxu1067** [comment]
@ksivaman, ok, merged #88 with this PR and close #88
.

File: docs/api/jax.rst [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+Dataclasses
+-------------
+.. autoapiclass:: transformer_engine.jax.ShardingResource(dp_resource=None, tp_resource=None)
+
+
+Modules
+-------
+
+.. autoapiclass:: transformer_engine.jax.LayerNorm(epsilon=1e-6, layernorm_type='layernorm', **kwargs)
+    :members: __call__
```
Comment (ksivaman) [inline docs/api/jax.rst:24]:
```suggestion
  :members: __call__
```

Comment (ksivaman) [reply docs/api/jax.rst:24]: consistency

File: transformer_engine/jax/sharding.py [hunk 2]
```diff
@@ -36,11 +36,11 @@ class ShardingResource:
     Parameters
     ----------
     dp_resource : str, default = None
-        axis name in Mesh used to shard batch along.
-        if it is None, then disabling data parallelism.
+        The axis name in Mesh used to shard batches along.
+        If it is None, then disabling data parallelism.
```
Comment (ksivaman) [inline transformer_engine/jax/sharding.py:40]:
```suggestion
        If it is None, then data parallelism is disabled.
```

File: transformer_engine/jax/sharding.py [hunk 3]
```diff
@@ -36,11 +36,11 @@ class ShardingResource:
...
     tp_resource : str, default = None
-        axis name in Mesh used to split model tensor along.
-        if it is None, then disabling tensor parallelism.
+        The axis name in Mesh used to split the hidden dimensions along.
+        If it is None, then disabling tensor parallelism.
```
Comment (ksivaman) [inline transformer_engine/jax/sharding.py:43]:
```suggestion
        If it is None, then tensor parallelism is disabled.
```

File: transformer_engine/jax/transformer.py [hunk 4]
```diff
@@ -185,53 +185,55 @@ class MultiHeadAttention(nn.Module):
...
-        the number of attention heads
+        The number of attention heads
     dropout_rate : float, default = 0.0
-        dropout probability for the dropout op during multi-head attention.
+        Dropout probability for the dropout op during multi-head attention.
     dropout_rng_name: str, default = 'dropout'
-        the key in given RNGs via flax.linen.Module.apply that
-        for generate Dropout masks in the core attention.
+        The key in given RNGs via flax.linen.Module.apply that
+        to generate Dropout masks in the core attention.
```
Comment (ksivaman) [inline transformer_engine/jax/transformer.py:195]: This doesn't read too well. Maybe just some grammar fix

Comment (ksivaman) [inline transformer_engine/jax/transformer.py:195]:
```suggestion
        The key in given RNGs via flax.linen.Module.apply that are
        used to generate Dropout masks in the core attention.
```

Comment (ksivaman) [reply transformer_engine/jax/transformer.py:195]: Is this what it means? @mingxu1067

Comment (mingxu1067) [reply transformer_engine/jax/transformer.py:195]: Yes, it is. Change to the suggestion. THX

File: transformer_engine/jax/transformer.py [hunk 5]
```diff
@@ -185,53 +185,55 @@ class MultiHeadAttention(nn.Module):
...
     layernorm_type : {'layernorm', 'rmsnorm'}, default = 'layernorm'
-        indicate the type of layer normalization.
+        Indicate the type of layer normalization.
     layernorm_epsilon: float, default = 1e-6
-        a value added to the denominator of layer normalization for numerical stability.
+        A value added to the denominator of layer normalization for numerical stability.
     kernel_init: Initializer, default =
         flax.linen.initializers.variance_scaling(1.0, 'fan_in', 'normal')
-        used for initializing weights of QKV and Output projection weights.
+        Used for initializing weights of QKV and Output projection weights.
```
Comment (ksivaman) [inline transformer_engine/jax/transformer.py:202]:
```suggestion
        Used for initializing the QKV and Output projection weights.
```

File: transformer_engine/jax/transformer.py [hunk 6]
```diff
@@ -185,53 +185,55 @@ class MultiHeadAttention(nn.Module):
...
+        It should be a callable object with three arguments (jax.random.PRNGKey, shape, dtype).
     use_bias: bool, default = False
-        indicate whether to enable bias shifting for QKVO projections.
-        if set to False, the layer will not learn additive biases.
+        Indicate whether to enable bias shifting for QKVO projections.
+        If set to False, the layer will not learn additive biases.
     bias_init: Initializer, default = flax.linen.initializers.zeros
-        used for initializing bias of QKVO projections, only works when :attr:`use_bias=True`.
+        Used for initializing bias of QKVO projections, it only works when :attr:`use_bias=True`.
```
Comment (ksivaman) [inline transformer_engine/jax/transformer.py:208]:
```suggestion
        Used for initializing bias of QKVO projections, only used when :attr:`use_bias=True`.
```

File: transformer_engine/jax/transformer.py [hunk 7]
```diff
@@ -185,53 +185,55 @@ class MultiHeadAttention(nn.Module):
...
+        A value added to the denominator of layer normalization for numerical stability.
     kernel_init: Initializer, default =
         flax.linen.initializers.variance_scaling(1.0, 'fan_in', 'normal')
-        used for initializing weights of QKV and Output projection weights.
+        Used for initializing weights of QKV and Output projection weights.
+        It should be a callable object with three arguments (jax.random.PRNGKey, shape, dtype).
     use_bias: bool, default = False
-        indicate whether to enable bias shifting for QKVO projections.
-        if set to False, the layer will not learn additive biases.
+        Indicate whether to enable bias shifting for QKVO projections.
```
Comment (ksivaman) [inline transformer_engine/jax/transformer.py:205]:
```suggestion
        Indicate whether or not to enable bias shifting for QKVO projections.
```

File: transformer_engine/jax/__init__.py [hunk 8]
```diff
@@ -5,5 +5,6 @@
 from .fp8 import fp8_autocast, update_collections, update_fp8_metas
 from .module import DenseGeneral, LayerNormDenseGeneral, LayerNormMLP, TransformerEngineBase
 from .transformer import extend_logical_axis_rules
-from .transformer import RelativePositionBiases, TransformerLayer, TransformerLayerType
+from .transformer import RelativePositionBiases, MultiHeadAttention
+from .transformer import TransformerLayer, TransformerLayerType
 from .sharding import ShardingResource
```
[the remark this answers is not in the document]
Comment (mingxu1067) [reply transformer_engine/jax/__init__.py]: Fixed

File: docs/api/jax.rst [hunk 9]
```diff
@@ -0,0 +1,53 @@
+..
+    Copyright (c) 2022, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
```
[the remark this answers is not in the document]
Comment (mingxu1067) [reply docs/api/jax.rst]: Fixed

File: docs/api/jax.rst (hunk 1 above)
Comment (mingxu1067) [reply docs/api/jax.rst]: Fixed

File: transformer_engine/jax/sharding.py (hunk 2 above)
Comment (mingxu1067) [reply transformer_engine/jax/sharding.py]: Fixed

File: transformer_engine/jax/sharding.py (hunk 3 above)
Comment (mingxu1067) [reply transformer_engine/jax/sharding.py]: Fixed

File: transformer_engine/jax/transformer.py (hunk 4 above)
Comment (mingxu1067) [reply transformer_engine/jax/transformer.py]: Fixed as the suggestion below

Comment (mingxu1067) [reply transformer_engine/jax/transformer.py]: Fixed

File: transformer_engine/jax/transformer.py (hunk 5 above)
Comment (mingxu1067) [reply transformer_engine/jax/transformer.py]: Fixed

File: transformer_engine/jax/transformer.py (hunk 6 above)
Comment (mingxu1067) [reply transformer_engine/jax/transformer.py]: Fixed

File: transformer_engine/jax/transformer.py (hunk 7 above)
Comment (mingxu1067) [reply transformer_engine/jax/transformer.py]: Fixed

File: transformer_engine/jax/fp8.py [hunk 10]
```diff
@@ -338,12 +338,11 @@ def fp8_autocast(enabled: bool = False,
...
 # Function Wrappers
-def update_collections(new: Collection, original: Collection) -> Collection:
+def update_collections(new: Collection, original: Collection) -> FrozenDict:
     r"""
-    A helper to update Flax's Collection. Collection is a union type of dict and
-    Flax's FrozenDict.
+    A helper to update Flax's Collection.
 
-    Collection = [dict, FrozenDict]
+    Collection = [dict, Flax's FrozenDict]
```
Comment (ksivaman) [inline transformer_engine/jax/fp8.py:345]:
```suggestion
    Collection = [dict, flax.core.frozen_dict.FrozenDict]
```

File: transformer_engine/jax/fp8.py [hunk 11]
```diff
@@ -364,14 +363,16 @@ def update_fp8_metas(state: Collection) -> Collection:
...
+
+        exp = floor(log2(fp8_max / amax)) - margin
+        sf = round(power(2, abs(exp)))
+        sf = sf if amax > 0.0, else original_scale
+        sf = sf if isfinite(amax), else original_scale)
+        updated_scale = 1/sf if exp < 0, else sf
+        updated_scale_inv = 1/updated_scale
 
-    Collection = [dict, FrozenDict]
+    Collection = [dict, Flax's FrozenDict]
```
Comment (ksivaman) [inline transformer_engine/jax/fp8.py:375]:
```suggestion
    Collection = [dict, flax.core.frozen_dict.FrozenDict]
```

File: transformer_engine/jax/module.py [hunk 12]
```diff
@@ -103,14 +103,14 @@ class Softmax(nn.Module):
     Parameters
     ----------
     scale_factor : float, default = 1.0
-        scale the inputs along the last dimension before running softmax.
+        Scale the inputs along the last dimension before running softmax.
```
Comment (ksivaman) [inline transformer_engine/jax/module.py:106]: Is this scaling only across the last dimension? The whole softmax input is scaled, right?

File: transformer_engine/jax/module.py [hunk 13]
```diff
@@ -426,56 +430,60 @@ class LayerNormDenseGeneral(TransformerEngineBase):
...
-        the name of axes used to shard the weights with a corresponding mesh.
+        The name of axes used to shard the weights with a corresponding mesh.
     use_bias: bool, default = False
-        indicate whether to enable bias shifting.
-        if set to False, the layer will not learn an additive bias.
+        Indicate whether to enable bias shifting.
+        If set to False, the layer will not learn an additive bias.
     bias_init: Initializer, default = flax.linen.initializers.zeros
-        used for initializing bias, only works when :attr:`use_bias=True`.
+        Used for initializing bias, only works when :attr:`use_bias=True`.
```
Comment (ksivaman) [inline transformer_engine/jax/module.py:463]:
```suggestion
        Used for initializing bias, only used when :attr:`use_bias=True`.
```

File: transformer_engine/jax/transformer.py [hunk 14]
```diff
@@ -612,56 +647,59 @@ class TransformerLayer(nn.Module):
...
     use_bias: bool, default = False
-        indicate whether to enable bias shifting for QKVO projections, FC1 and FC2.
-        if set to False, the layer will not learn additive biases.
+        Indicate whether to enable bias shifting for QKVO projections, FC1 and FC2.
+        If set to False, the layer will not learn additive biases.
     bias_init: Initializer, default = flax.linen.initializers.zeros
-        used for initializing bias of QKVO projections,
-        FC1 and FC2, only works when :attr:`use_bias=True`.
+        Used for initializing bias of QKVO projections,
+        FC1 and FC2. It only works when :attr:`use_bias=True`.
```
Comment (ksivaman) [inline transformer_engine/jax/transformer.py:684]:
```suggestion
        FC1 and FC2. It only used when :attr:`use_bias=True`.
```

Comment (ksivaman) [reply transformer_engine/jax/transformer.py:684]: @mingxu1067 This pattern actually exists throughout, could you please go through all the cases here change "only works" -> "only used" so that it gives the correct picture?

File: transformer_engine/jax/module.py (hunk 12 above)
Comment (mingxu1067) [reply transformer_engine/jax/module.py:106]: Yes, it is.

File: transformer_engine/jax/module.py [hunk 15]
```diff
@@ -98,19 +98,25 @@ def _combine_biases(*masks: List[Array]):
...
+        softmax_mask = mask * -1e-10
+        output = softmax(masked_scaled + softmax_mask)
 
     Parameters
     ----------
     scale_factor : float, default = 1.0
-        scale the inputs along the last dimension before running softmax.
-    softmax_type : SoftmaxType, default = 'layernorm'
-        indicate the type of softmax.
+        Scale the whole (inputs + bias) before running softmax.
```
Comment (ksivaman) [inline transformer_engine/jax/module.py:112]:
```suggestion
        Scalar for the input to softmax.
```

File: docs/api/jax.rst [hunk 16]
```diff
@@ -0,0 +1,53 @@
+..
+    Copyright (c) 2022-2023, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+
+    See LICENSE for license information.
+
+Jax
+=======
+
+Types
```
Comment (ksivaman) [inline docs/api/jax.rst:9]:
```suggestion
Enums
```

File: docs/api/jax.rst [hunk 17]
```diff
@@ -0,0 +1,53 @@
...
+  :members: __call__
+
+.. autoapiclass:: transformer_engine.jax.RelativePositionBiases(num_buckets, max_distance, num_heads, **kwargs)
+  :members: __call__
+
+.. autoapiclass:: transformer_engine.jax.MultiHeadAttention(head_dim, num_heads, **kwargs)
+  :members: __call__
+
+.. autoapiclass:: transformer_engine.jax.TransformerLayer(hidden_size=512, mlp_hidden_size=2048, num_attention_heads=8, **kwargs)
+  :members: __call__
```
Comment (ksivaman) [inline docs/api/jax.rst:42]: All of these modules are showing up in the `Functions` category below in the generated docs. @mingxu1067

File: docs/api/framework.rst [hunk 18]
```diff
@@ -9,3 +9,4 @@ Framework-specific API
 .. toctree::
 
     pytorch
+    jax
```
Comment (nouiz) [inline docs/api/framework.rst:12]: The main README should also mention JAX.
Grep for pytorch in that document and add JAX at those places and add a JAX example too as there is a PyTorch example.

File: transformer_engine/jax/module.py (hunk 15 above)
Comment (mingxu1067) [reply transformer_engine/jax/module.py]: Fixed

File: docs/api/jax.rst (hunk 16 above)
Comment (mingxu1067) [reply docs/api/jax.rst]: Remove categoried as PyTorch to make the style consistent.

File: docs/api/jax.rst (hunk 17 above)
Comment (mingxu1067) [reply docs/api/jax.rst:36]: Remove categoried as PyTorch to make the style consistent.

File: docs/api/framework.rst (hunk 18 above)
Comment (mingxu1067) [reply docs/api/framework.rst:12]: I think README.md need more refactor to show all FWs supported by TE. Diretort adding JAX to all Pytorch appearance might mess up reading. We can submit a split PR to refactor README.md.

**Review by ksivaman — APPROVED** [review]
